Getroles logroles

michaellasky opened the pull request:
<!--
Thanks for your interest in the project. Bugs filed and PRs submitted are appreciated!

Please make sure that you are familiar with and follow the Code of Conduct for
this project (found in the CODE_OF_CONDUCT.md file).

Also, please make sure you're familiar with and follow the instructions in the
contributing guidelines (found in the CONTRIBUTING.md file).

If you're new to contributing to open source projects, you might find this free
video course helpful: http://kcd.im/pull-request

Please fill out the information below to expedite the review and (hopefully)
merge of your pull request!
-->

<!-- What changes are being made? (What feature/bug is being fixed here?) -->
This is a draft implementation for the functionality requested in #282  

The intention is to give a jumping off point for discussion.  
 
<!-- How were these changes implemented? -->

I've added a file named role-helpers.js which implements the two requested functions, and also moved the function buildElementRoleList from src/queries/role.js into role-helpers.js.  Since role-helpers needed similar functionality, and it didn't seem to make sense to export it from query/role with a bunch of query functions.  

I also added a dependency on lodash.merge for a deep merge in getRoles.  Maybe this could be removed because the merge happening there isn't all that complicated.  But using the premade merge() provided by lodash allowed me to keep the initial PR code cleaner for discussion.  

<!-- Have you done all of these things?  -->

**Checklist**:

<!-- add "N/A" to the end of each line that's irrelevant to your changes -->

<!-- to check an item, place an "x" in the box like so: "- [x] Documentation" -->

- [x] Documentation added to the
      [docs site](https://github.com/alexkrolick/testing-library-docs)
- [x] Typescript definitions updated
- [x] Tests
- [x] Ready to be merged
      <!-- In your opinion, is this ready to be merged as soon as it's reviewed? -->

I'm still pretty new to this particular code base, so I don't think  it's ready to be merged without some significant review and iteration from more experienced members.  

<!-- feel free to add additional comments -->

michaellasky commented:
I have code for a test that doesn't rely on snapshotting if that's easier for the CI.  I'm using a snapshot because the test shouldn't necessarily fail (beyond needing to update the snapshot) if the formatting from debugDOM changes.  If I hardcode the output from debugDOM into a test case, then the code would need to be updated if debugDOM formatting changes.

kentcdodds commented on src/__tests__/role-helpers.js:
```diff
@@ -0,0 +1,120 @@
+  })
+})
+
+test('logRoles is a function', () => {
```
I don't think we need this test as the other test calls this function.

kentcdodds commented on src/__tests__/role-helpers.js:
```diff
@@ -0,0 +1,120 @@
+    input2,
+  }
+}
+test('getRoles is a function', () => {
```
I don't think we need this test.

michaellasky replied:
I don't disagree, but just for discussion the reason I include a test like that is that it distinguishes between problems of implementation and problems with configuration/plumbing.  If that test passes and the others fail then you know it's a problem with the implementation.  If that one fails along with the others then it lets you know at a glance there's some other problem (it's not being exported or something).  

But in this, and many cases it doesn't matter, and is mostly a way to get past the 'blank page' syndrome, a test to get started, haha.  

I'll pull them out.

kentcdodds commented on src/__tests__/role-helpers.js:
```diff
@@ -0,0 +1,120 @@
+import {getRoles, logRoles} from '../role-helpers'
+
+function setup() {
```
Maybe we could use this instead:

```javascript
const section = document.createElement('section')
section.innerHTML = `
  the html for everything else...
`
```

The way it is currently is a bit hard to visualize.

michaellasky replied:
Will do.  The only sticking point with doing it that way is that I then have to reach into that dom and pull all those elements out to add them to my expected obj.  But I think the readability of the dom itself is a bigger win, so i agree something like your suggestion should be done.

michaellasky commented:
> Also, a screenshot would be helpful!


[Quick screenshot from npm test with a console log](https://drive.google.com/file/d/1xpzwVwUqQzQFqqvGKg1U2CB7FXr-OgXL/view?usp=sharing)  I'll spin it up in the browser in a bit when I make the other changes and take a look at / screencap the output there.  

I can of course make any changes to the formatting.  I didn't match the suggested implementation exactly, but can if that's wanted.

kentcdodds commented:
Looks pretty good to me!

1. It would probably be helpful to indent the DOM output so the sections are more visible
2. It would probably be helpful to have a separator (even just a single line) between each DOM output so people don't think they represent the same node or something.
3. That `<article>` node has a bunch of stuff in it and if it were really big, that output could be VERY extensive. I wonder if we should just clip it to `<article />` instead of `<article>all the contents</article>` Thoughts?

michaellasky commented:
>     3. That `<article>` node has a bunch of stuff in it and if it were really big, that output could be VERY extensive. I wonder if we should just clip it to `<article />` instead of `<article>all the contents</article>` Thoughts?

I actually had a similar thought.  I didn't pursue it because that would be more of a change to debugDOM I think?  We'd need to add a parameter to tell it not to print all the children (assuming it can't already do that, i haven't looked at it extensively).  The current implementation is just debugDom(node), which in the case of a large node like the article node, prints out all the children.  

In any event, i don't think it'd be very difficult to handle.

kentcdodds commented:
My thinking is that before we hand it to `debugDOM` we could create a copy of the DOM node (via something similar to [asFragment](https://github.com/testing-library/react-testing-library/blob/3076b119172b8b7188603509bfe576b7d2914de3/src/index.js#L69-L80)) and then remove all the children.

michaellasky commented:
Oh, I hadn't considered that.  That's not a bad idea, I'll give it a shot and update this (hopefully tonight, time permitting) with that and the other suggested changes.

michaellasky commented:
So I've made most of the discussed changes.

[This is what the output of logRoles looks like now](https://drive.google.com/file/d/1AXMSkON1U6Xemj8eL2I7RGhGpZmUE6yB/view?usp=sharing)

I also fixed a bug with getRoles, and refactored it to clean it up and simplify things.  

logRoles uses a snapshot test again.  The regex only worked if the html under test is very simple, and it a tad more 'real world'.  So rather than putting an ugly regex in the test i went back to a snapshot.  I'm not sure what is needed to tell travis to save the snapshot so it'll pass the build though.  

Lodash.merge dependency is gone.

michaellasky commented:
I'm pretty sure the current build failure is unrelated to these code changes.  

wait-for-element-to-be-removed.js is timing out at some point in one of the builds, which I had noticed happening intermittently when i had issues earlier (when my build was also failing for coverage reasons).

The assumption is that it'd (probably) pass if i closed and re-opened the PR to trigger a rebuild, but I'm leaving it so someone who knows more about what's going on can verify.

kentcdodds commented on src/__tests__/__snapshots__/role-helpers.js.snap:
```diff
@@ -37,6 +37,28 @@ article ###################################
 
+command ###################################
+
+[36m<menuitem[39m
```
Maybe we could use https://github.com/joaogranado/jest-serializer-ansi to make these snapshots better.

michaellasky replied:
Added jest-serilizer-ansi  

Also updated a couple other tests ([e452465](https://github.com/testing-library/dom-testing-library/pull/285/commits/e4524653e4b0212063d4d2e715cf99bdb479b7c2)) that had inline snapshots containing the same control characters.

kentcdodds commented on src/__tests__/__snapshots__/role-helpers.js.snap:
```diff
@@ -0,0 +1,166 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
+
+exports[`logRoles logs expected roles for various dom nodes 1`] = `
+"region ####################################
```
Rather than `##################` could we use `----------------------`?

michaellasky replied:
Switched [deb5bf5c](https://github.com/testing-library/dom-testing-library/pull/285/commits/de5bf5c2a184ed999e80e33b7ec3f05af767b22c)

kentcdodds commented on src/__tests__/__snapshots__/role-helpers.js.snap:
```diff
@@ -0,0 +1,166 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
+
+exports[`logRoles logs expected roles for various dom nodes 1`] = `
```
I think this looks really good. Would love to hear what others have to say though.

afontcu commented on src/index.js:
```diff
@@ -12,6 +12,7 @@ export * from './get-node-text'
 export * from './events'
 export * from './get-queries-for-element'
 export * from './query-helpers'
+export * from './role-helpers'
```
I love this PR! I learned a ton by reading your code and comments. 

Now that you've discussed public API, I'm still not sold on the idea of the current naming. I feel that, as exposed in https://github.com/testing-library/dom-testing-library/issues/282#issuecomment-502298508, `getRoles` could easily be confused with one of the `getBy*` queries from DTL. What do you folks think about that?

kentcdodds replied:
I don't think all that many people do this:

```javascript
import {getByRole} from '@testing-library/dom'
```

I'm pretty sure most people do this:

```javascript
import {render} from '@testing-library/some-implementation'

const {getByRole} = render(myThing)
```

Or this:

```javascript
import {getQueriesForElement} from '@testing-library/dom'

const {getByRole} = getQueriesForElement(someContainer)
```

So I'm not very concerned about this.

michaellasky replied:
Good catch on the exports, I had completely forgotten to revisit those.  

I've removed the elementRoleList export, that is certainly not needed anymore, now that getImplicitAriaRoles is moved from src/queries/role to src/role-helpers.

We do need to export getImplicitAriaRoles  from role-helpers because src/queries/role uses it.  That said, I don't think it really needs to be part of the public API.  

Should we just export getRoles and logRoles in index?  Will commit that change after hearing back.

kentcdodds replied:
Yeah, I think that's right 👍

michaellasky replied:
[Done b9a16e1](https://github.com/testing-library/dom-testing-library/pull/285/commits/b9a16e1f27d4f57e53d16b992563f4b1d4097303)

I'll get started on documentation tonight or tomorrow as well, now that things are starting to come together.

michaellasky commented on src/__tests__/role-helpers.js:
```diff
@@ -0,0 +1,182 @@
+  const input2 = getByTestId('a-input-2')
+  const textarea = getByTestId('a-textarea')
+
+  return {
```
Great idea, done.  [2ecda86](https://github.com/testing-library/dom-testing-library/pull/285/commits/2ecda860af4cfa0bb5e22332188b1f0e14cc7dbe)

michaellasky commented:
I've attempted to add the typescript definitions but I seem to be doing something wrong.  I'm not particularly fluent with typescript. 

I'd imagine the definitions, in a file named typings/role-helpers.d.ts, should look something like:
  
    export function logRoles(container: HTMLElement): string
    export function getRoles(
        container: HTMLElement,
     ): {[index: string]: HTMLElement[]}

But unfortunately that fails linting (kcd-scripts) on build with:

2:1  error  Parsing error: Unexpected token, expected "{"

After some experimentation, I can only get it to pass if I add an implementation for some reason.  For debugging purposes, I changed it to:

    export function logRoles(container: HTMLElement): string
    export function getRoles(container: HTMLElement): string

Which fails in the same way.  But:

    export function logRoles(container: HTMLElement): string { return container? '1': '2' }
    export function getRoles(container: HTMLElement): string { return container? '1': '2' }

passes linting fine.  I am confuse.  But it's probably just some assumption I'm making because of lack of TS experience.

kentcdodds commented:
Yeah, honestly I'm not sure what's up with that 😞 we should probably just disable linting on that file. This always happens to me when I change that file so I just commit that file with --no-verify 😬

I think things look pretty good. I'd love another review though!

kentcdodds commented:
Oh, and it looks like there's a merge conflict 😬

michaellasky commented:
Added the typescript definitions.  Whoever else reviews these changes should pay particular attention to them, since I don't have complete confidence they're correct.  

[93f9530](https://github.com/testing-library/dom-testing-library/pull/285/commits/93f9530baf31ecdb848044f6272d143744107cdd

alexkrolick commented on src/role-helpers.js:
```diff
@@ -0,0 +1,90 @@
+
+  return Object.entries(roles)
+    .map(([role, elements]) => {
+      const numDelimeters = 42 - role.length // 42 is arbitrary
```
As far as I know there aren't any aria-roles greater than 42 chars, but I would really hate for a formatting error to break someone's test. Can we be conservative here and do a Math.min(0, numDelimiters) or move the delimiter characters to a fixed length line?

michaellasky replied:
That's a good point.  I think it's very unlikely there will ever be a role greater than 42 characters long... there's also no real downside to protecting against that.  Maybe I was being too clever with embedding the name in the delimiters.  It read easier (to my eyes) with just a single heading/spacer element, but it probably makes more sense to just implement it like Kent originally had it and avoid any confusion, extra code, or bugs (however unlikely). 

I'll make that change this evening unless I hear an objection.

michaellasky replied:
Done [593c33f](https://github.com/testing-library/dom-testing-library/pull/285/commits/593c33f987bba473a3c0c89f446e8a1af604d7fd)

Check the snapshot for output format.  

I would have like to indent the elements themselves to further distinguish them from the headings, but unfortnately I couldn't find a decent way to do that with the way debugDOM->prettyDOM->prettyFormat works.  They inject their own newlines into the output, so if I try to indent it I can only indent the first line.  I'd need to split the output of debugDOM up by lines and indent them individually... which seemed messy for a minor visual improvement.  

I also purposefully left out the space between the delimeter bar ('-----') and the role name label.  With the space it looks a lot louder visually. 

     --------------------------------------------------
     navigation:
    <nav
        data-testid="a-nav"
    />

    --------------------------------------------------
    heading:

    <h1
      data-testid="a-h1"
    />

vs

     --------------------------------------------------

     navigation:
    <nav
        data-testid="a-nav"
    />

    --------------------------------------------------

    heading:

    <h1
      data-testid="a-h1"
    />

kentcdodds commented:
@all-contributors please add @michaellasky for code, tests, and docs

kentcdodds commented:
:tada: This PR is included in version 5.3.0 :tada:

The release is available on:
- `npm package (@latest dist-tag)`
- [GitHub release](https://github.com/testing-library/dom-testing-library/releases/tag/v5.3.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:

kentcdodds commented:
Thanks so much for your help! I've added you as a collaborator on the project. Please make sure that you review the `other/MAINTAINING.md` and `CONTRIBUTING.md` files (specifically the bit about the commit messages and the git hooks) and familiarize yourself with the code of conduct (we're using [the contributor covenant](http://contributor-covenant.org/)). You might also want to [watch the repo](https://help.github.com/articles/watching-repositories/) to be notified when someone files an issue/PR. Please continue to make PRs as you feel the need (you can make your branches directly on the repo rather than your fork if you want). Thanks! And welcome to the team :)

michaellasky commented:
Thanks Kent, you're making open source a lot of fun.  I'll continue to watch and contribute.
